src/webgpu/api/operation/memory_sync/buffer/buffer_sync_test.ts

```diff
@@ -0,0 +1,242 @@
+  createStorageWriteRenderPipeline(value: number): GPURenderPipeline {
+    const wgslVertex = `
+      [[builtin(position)]] var<out> Position : vec4<f32>;
+      [[stage(vertex)]] fn vert_main() -> void {
+        Position = vec4<f32>(0.5, 0.5, 0.0, 1.0);
+        return;
+      }
+    `;
+
+    const wgslFragment = `
+      [[location(0)]] var<out> outColor : vec4<f32>;
+      type Data = [[block]] struct {
+        [[offset(0)]] a : i32;
+      };
+
+      [[binding(0), set(0)]] var<storage_buffer> data : Data;
+      [[stage(fragment)]] fn frag_main() -> void {
+        data.a = ${value};
+        outColor = vec4<f32>(1.0, 0.0, 0.0, 1.0);
+        return;
+      }
+    `;
+
+    return this.device.createRenderPipeline({
+      vertexStage: {
+        module: this.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'vert_main',
+      },
+      fragmentStage: {
+        module: this.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'frag_main',
+      },
+      primitiveTopology: 'point-list',
+      colorStates: [{ format: 'rgba8unorm' }],
+    });
+  }
+
+  beginSimpleRenderPass(encoder: GPUCommandEncoder): GPURenderPassEncoder {
+    const view = this.device
+      .createTexture({
+        size: { width: 4, height: 4, depth: 1 },
+        format: 'rgba8unorm',
+        usage: GPUTextureUsage.OUTPUT_ATTACHMENT,
+      })
+      .createView();
+    return encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: view,
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+        },
+      ],
+    });
+  }
+
+  // Write buffer via draw call in render pass. Use bundle if needed.
+  writeByRenderPass(bundle: boolean, buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteRenderPipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+
+    const pass = this.beginSimpleRenderPass(encoder);
+    const renderer = bundle
+      ? this.device.createRenderBundleEncoder({ colorFormats: ['rgba8unorm'] })
+      : pass;
+    renderer.setBindGroup(0, bindGroup);
+    renderer.setPipeline(pipeline);
+    renderer.draw(1, 1, 0, 0);
+
+    if (bundle) pass.executeBundles([(renderer as GPURenderBundleEncoder).finish()]);
+    pass.endPass();
+  }
+
+  // Write buffer via dispatch call in compute pass.
+  writeByComputePass(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
+    const pipeline = this.createStorageWriteComputePipeline(value);
+    const bindGroup = this.createBindGroup(pipeline, buffer);
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatch(1);
+    pass.endPass();
+  }
+
+  // Write bufer via BuferToBuffer copy.
+  async writeByB2BCopy(buffer: GPUBuffer, value: number, encoder: GPUCommandEncoder) {
```

Comment:
nit: encodeWriteByB2BCopy, and so on below.

It's really important that these encode writes, not that they actually write. In particular, for the async ones,f they actually wrote immediately, then the async tmpBuffer initialization would happen inline instead of ahead of time. In addition to the naming change, maybe add a comment clarifying this.

Reply:
Done. I added a comment as you suggested, but I am not 100% sure whether it is appropriate, though.

Reply:
Wording looks good, but actually please include it in the comment on each of these functions, and change the comment style to a `/** doc comment */` so it will be picked up by IDEs.

Also nit: bufer -> buffer